workbench-core/accounts/src/services/projectService.ts

```diff
@@ -127,91 +141,58 @@ export default class ProjectService {
-      key: { name: 'resourceType', value: this._resourceType },
-      index: 'getResourceByCreatedAt',
-      limit: pageSize,
-      filter: filterExpression,
-      values: expressionAttributeValues
-    };
-
-    let projects: Project[] = [];
-
-    let returnedCount = 0;
-    while (returnedCount < pageSize && returnedCount < projectIds.length) {
-      queryParams = addPaginationToken(paginationToken, queryParams);
-
-      const projectsResponse = await this._aws.helpers.ddb.query(queryParams).execute();
-
-      returnedCount += projectsResponse.Count!;
-
-      // nothing returned
-      if (projectsResponse.Items === undefined || _.isEmpty(projectsResponse.Items)) {
-        paginationToken = getPaginationToken(projectsResponse);
-        console.log(paginationToken);
-      }
-
-      // nothing else to get from DDB
-      else if (projectsResponse.LastEvaluatedKey === undefined) {
-        projects = projects.concat(projectsResponse.Items.map((item) => this._formatFromDDB(item)));
-        paginationToken = undefined;
+    const keys: { [key: string]: unknown }[] = projectIds.map((projectId) =>
+      buildDynamoDBPkSk(projectId, resourceTypeToKey.project)
+    );
+    const projectsResponse = (await this._aws.helpers.ddb.get(keys).execute()) as BatchGetItemCommandOutput;
+    if (projectsResponse.Responses) {
+      const projects: Project[] = projectsResponse.Responses[this._tableName].map((item) =>
+        this._formatFromDDB(item)
+      );
+      // apply sort or filter
+      let projectsOnPage: Project[] = projects;
+      if (filter) {
+        projectsOnPage = manualFilterProjects(filter, projects);
       }
-
-      // too little--while loop continues
-      else if (returnedCount < pageSize) {
-        projects = projects.concat(projectsResponse.Items.map((item) => this._formatFromDDB(item)));
-        paginationToken = getPaginationToken(projectsResponse);
+      if (sort) {
+        projectsOnPage = manualSortProjects(sort, projects);
+      } else if (filter === undefined) {
+        // default sort is by createdAt
+        projectsOnPage = manualSortProjects({ createdAt: 'asc' }, projects);
       }
-
-      // juuuust the right amount--getPaginationToken from response and exit loop
-      else if (returnedCount === pageSize) {
-        paginationToken = getPaginationToken(projectsResponse);
-        projects = projects.concat(projectsResponse.Items.map((item) => this._formatFromDDB(item)));
+      // build page and pagination token
+      if (paginationToken) {
+        const manualExclusiveStartKey = fromPaginationToken(paginationToken);
+        const exclusiveStartProjectId = manualExclusiveStartKey.pk.split('#')[1];
+        const exclusiveStartProject = projectsOnPage.find(
+          (project) => project.id === exclusiveStartProjectId
+        );
+        if (exclusiveStartProject === undefined) {
+          throw Boom.badImplementation('Something went wrong with listing projects');
```

Comment:
So, do you think `badRequest` is better? The error could be due to an invalid pagination token or just something going wrong in finding the id from the pagination token, so I didn't know which one was best.

Reply:
Because the pagination token is being sent in by the user, it should be a 4xx: probably a 422 since it is syntactically correct but logically unprocessable, but it looks like we've been going with 400s for validation errors.

Reply:
synced with tyler--changing to `badRequest`